v0/customerio/transform.js

```diff
@@ -0,0 +1,222 @@
+
+// Get the spec'd traits, for now only address needs treatment as 2 layers.
+// populate the list of spec'd traits in constants.js
+const populateSpecedTraits = (payload, message) => {
+  //console.log(message);
+  SpecedTraits.forEach(trait => {
+    const mapping = TraitsMapping[trait];
+    const keys = Object.keys(mapping);
+    keys.forEach(key => {
+      set(payload, key, get(message, mapping[key]));
+    });
+  });
+};
+
+function responseBuilder(message, evType, evName, destination) {
+  const rawPayload = {};
+  let endpoint;
+  let requestConfig = defaultPostRequestConfig;
+  //console.log(message);
+  let userId =
+    message.userId && message.userId != "" ? message.userId : undefined;
+
+  const response = defaultRequestConfig();
+  response.userId = message.userId ? message.userId : message.anonymousId;
+
+  if (evType === EventType.IDENTIFY) {
+    // populate speced traits
+    populateSpecedTraits(rawPayload, message);
+    if (message.context.traits) {
+      const traits = Object.keys(message.context.traits);
+      traits.forEach(trait => {
+        //populate keys other than speced traits
+        if (!SpecedTraits.includes(trait)) {
+          set(rawPayload, trait, get(message, "context.traits." + trait));
+        }
+      });
+    }
+
+    if (message.user_properties) {
+      const userProps = Object.keys(message.user_properties);
+      userProps.forEach(prop => {
+        let val = get(message, "user_properties." + prop);
+        // send only top level keys
+        if (isPrimitive(val)) {
+          set(rawPayload, prop, val);
+        }
+      });
+    }
+
+    set(
+      rawPayload,
+      "created_at",
+      Math.floor(new Date(message.originalTimestamp).getTime() / 1000)
+    );
+    //console.log(rawPayload);
+
+    if (userId) {
+      endpoint = IDENTITY_ENDPOINT.replace(":id", userId);
+    } else {
+      return { statusCode: 400, error: "userId not present" };
+    }
+    requestConfig = defaultPutRequestConfig;
+  } else {
+    if (message.properties) {
+      // use this if only top level keys are to be sent
+
+      /* const eventProps = Object.keys(message.properties);
+      eventProps.forEach(prop => {
+        let val = get(message, "properties." + prop);
+        if (isPrimitive(val)) {
+          set(rawPayload.data, prop, val);
+        }
+      }); */
+
+      if (deviceDeleteRelatedEventName == evName) {
+        const token = get(message, "context.device.token");
+        if (userId && token) {
+          endpoint = DEVICE_DELETE_ENDPOINT.replace(":id", userId).replace(
+            ":device_id",
+            token
+          );
+
+          response.endpoint = endpoint;
+          response.method = "DELETE";
+          response.headers = {
+            Authorization:
+              "Basic " +
+              btoa(destination.Config.siteID + ":" + destination.Config.apiKey)
+          };
+
+          return response;
+        } else {
```

Comment:
Else can be skipped. Anyways, it's better to return error in the if.
I like this pattern.

https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88

Reply:
The post is really good. I'd try to enforce this pattern going forward